File: python/isbn-verifier/isbn_verifier.py
"""Exercism's ISBN Verifier challenge."""


def is_valid(isbn: str) -> bool:
    """Return True if the given ISBN-10 string is valid.

    :param isbn: str - given ISBN-10 string, may be hyphenated.
    :return: bool - True if ISBN-10 is valid, False otherwise."""
    parsed = list(isbn.replace("-", ""))
    # ISBN must be 10 digits
    if len(parsed) != 10:
        return False
    # Validate the check digit
    if parsed[-1].lower() == "x":
        parsed[-1] = "10"
    # Discard ISBN if contains non-digits
    if not all((char.isdigit() for char in parsed)):
        return False
    # Calculate by the ISBN verification formula
    total = sum((int(digit) * (10 - index)
                for index, digit in enumerate(parsed)))
    return total % 11 == 0

File: go/isbn-verifier/isbn_verifier.go
package isbnverifier

func IsValidISBN(isbn string) bool {
	position := 1
	sum := 0
	for index := len(isbn) - 1; index >= 0; index-- {
		char := isbn[index]
		digit := 0

		// hyphens are tolerated
		if char == '-' {
			continue
		}

		if position == 1 && char == 'X' {
			// X is only tolerated as the last digit
			digit = 10
		} else if char >= '0' && char <= '9' {
			digit = int(char - '0')
		} else {
			// non-digits are not tolerated
			return false
		}

		sum += position * digit
		position++
	}
	return position == 11 && sum%11 == 0
}
